File: optimizer/qp_solver.py
"""
blaque_baux/optimizer/qp_solver.py
────────────────────────────────────
Quadratic program portfolio optimizer using CVXPY.

This is exactly the Solver workflow from your finance grad school,
productionized for 15-minute rebalancing.

Problem formulation:
    minimize    w' Σ w  (portfolio variance)
    subject to:
        w' μ ≥ target_return          (minimum return constraint)
        sum(w_long)  = long_ratio     (long book sums to 1)
        sum(w_short) = -(1-long_ratio) (short book sums to -1)
        w_i ≥ min_wt  for i in longs  (minimum position)
        w_i ≤ max_wt  for i in longs  (maximum position)
        w_i ≤ -min_wt for i in shorts
        w_i ≥ -max_wt for i in shorts
        turnover penalty (minimize |w_new - w_prev|)

The optimizer also respects borrow costs by adjusting expected
returns for short positions before solving.

Outputs: weights dict {ticker: weight}
  Positive weights = long positions
  Negative weights = short positions
"""

import logging
import warnings
from typing import Dict, List, Optional, Tuple

import numpy as np
import pandas as pd
import cvxpy as cp

from config import optimizer_cfg

logger = logging.getLogger(__name__)

# Suppress CVXPY solver verbosity
warnings.filterwarnings("ignore", category=UserWarning)


# ── COVARIANCE ESTIMATION ─────────────────────────────────────────────────────

def ewma_covariance(
    returns: pd.DataFrame,
    lam: float = 0.94,
    min_periods: int = 20,
) -> np.ndarray:
    """
    Compute EWMA (exponentially weighted) covariance matrix.
    RiskMetrics standard: lambda = 0.94 for daily, 0.97 for monthly.
    For 15-min bars we use 0.94 (same decay, more observations).

    Returns: (n_assets × n_assets) covariance matrix
    """
    n = len(returns.columns)
    if len(returns) < min_periods:
        return np.eye(n) * returns.std().values ** 2  # diagonal fallback

    # Use pandas EWMA covariance
    ewm_cov = returns.ewm(com=(1 - lam) / lam, min_periods=min_periods).cov()

    # Extract the last timestamp's covariance matrix
    last_ts = ewm_cov.index.get_level_values(0)[-1]
    cov_matrix = ewm_cov.loc[last_ts].values

    # Ensure positive semi-definite (numerical cleanup)
    cov_matrix = _ensure_psd(cov_matrix)
    return cov_matrix


def _ensure_psd(matrix: np.ndarray, epsilon: float = 1e-6) -> np.ndarray:
    """
    Ensure matrix is positive semi-definite by clipping negative eigenvalues.
    """
    eigvals, eigvecs = np.linalg.eigh(matrix)
    eigvals = np.maximum(eigvals, epsilon)
    return eigvecs @ np.diag(eigvals) @ eigvecs.T


# ── MAIN OPTIMIZER ────────────────────────────────────────────────────────────

def optimize_portfolio(
    factor_scores: pd.Series,
    returns_history: pd.DataFrame,
    long_tickers: List[str],
    short_tickers: List[str],
    prev_weights: Optional[Dict[str, float]] = None,
    borrow_costs: Optional[Dict[str, float]] = None,
    confidence_levels: Optional[List[float]] = None,
) -> Tuple[Dict[str, float], Dict]:
    """
    Main portfolio optimizer. Runs the QP solve for one window.

    Args:
        factor_scores:     Series {ticker: score} — expected relative returns
        returns_history:   DataFrame of recent returns for covariance estimation
        long_tickers:      tickers selected for long book
        short_tickers:     tickers selected for short book
        prev_weights:      previous window's weights (for turnover penalty)
        borrow_costs:      annualized borrow cost per ticker (for HTB stocks)
        confidence_levels: list of confidence levels for efficient frontier
                           (mimics the Solver multi-confidence run from grad school)

    Returns:
        weights:    {ticker: weight}  (positive=long, negative=short)
        meta:       diagnostic info (expected_return, expected_vol, solve_status)
    """
    all_tickers = long_tickers + short_tickers
    n_long  = len(long_tickers)
    n_short = len(short_tickers)
    n_total = n_long + n_short

    if n_total < 4:
        logger.warning("Too few tickers for optimization, returning equal weight")
        return _equal_weight_fallback(long_tickers, short_tickers), {"status": "fallback"}

    # ── Expected returns (factor scores → per-window alpha estimates) ──────────
    mu = np.array([
        factor_scores.get(t, 0) * optimizer_cfg.target_return
        for t in all_tickers
    ])

    # Adjust for borrow costs on short positions
    if borrow_costs:
        for i, ticker in enumerate(all_tickers[n_long:], start=n_long):
            cost_annual = borrow_costs.get(ticker, 0)
            cost_per_window = cost_annual / (252 * 26)  # 26 windows/day approx
            mu[i] -= cost_per_window

    # ── Covariance matrix ───────────────────────────────────────────────────
    valid_hist = returns_history[all_tickers].dropna()
    if len(valid_hist) < 20:
        valid_hist = returns_history[all_tickers].fillna(0)

    cov = ewma_covariance(valid_hist)

    # ── CVXPY Variables ─────────────────────────────────────────────────────
    w = cp.Variable(n_total)

    # ── Objective: minimize variance + turnover penalty ────────────────────
    portfolio_variance = cp.quad_form(w, cov)

    if prev_weights is not None:
        w_prev = np.array([prev_weights.get(t, 0) for t in all_tickers])
        turnover = cp.norm1(w - w_prev)
        objective = cp.Minimize(
            optimizer_cfg.risk_aversion * portfolio_variance +
            optimizer_cfg.turnover_penalty * turnover
        )
    else:
        objective = cp.Minimize(optimizer_cfg.risk_aversion * portfolio_variance)

    # ── Constraints ─────────────────────────────────────────────────────────
    long_ratio  = optimizer_cfg.long_ratio
    short_ratio = 1 - long_ratio

    constraints = [
        # Long book: weights sum to long_ratio, each bounded
        cp.sum(w[:n_long]) == long_ratio,
        w[:n_long] >= optimizer_cfg.min_position_wt,
        w[:n_long] <= optimizer_cfg.max_position_wt,

        # Short book: weights sum to -short_ratio, each bounded
        cp.sum(w[n_long:]) == -short_ratio,
        w[n_long:] <= -optimizer_cfg.min_position_wt,
        w[n_long:] >= -optimizer_cfg.max_position_wt,

        # Minimum expected return (net of drag)
        mu @ w >= optimizer_cfg.target_return - optimizer_cfg.execution_drag * 2,
    ]

    # ── Solve ────────────────────────────────────────────────────────────────
    problem = cp.Problem(objective, constraints)

    try:
        problem.solve(solver=cp.CLARABEL, verbose=False)
    except Exception:
        try:
            problem.solve(solver=cp.SCS, verbose=False)
        except Exception as e:
            logger.warning(f"QP solve failed: {e}, using equal weight fallback")
            return _equal_weight_fallback(long_tickers, short_tickers), {"status": "failed"}

    if w.value is None or problem.status not in ["optimal", "optimal_inaccurate"]:
        return _equal_weight_fallback(long_tickers, short_tickers), {"status": problem.status}

    # ── Package results ──────────────────────────────────────────────────────
    weights = {t: float(w.value[i]) for i, t in enumerate(all_tickers)}

    # Clean up tiny weights (numerical noise)
    weights = {
        t: v for t, v in weights.items()
        if abs(v) >= optimizer_cfg.min_position_wt * 0.5
    }

    # Compute diagnostics
    w_arr = np.array([weights.get(t, 0) for t in all_tickers])
    expected_return = float(mu @ w_arr)
    expected_vol    = float(np.sqrt(w_arr @ cov @ w_arr))

    meta = {
        "status":          problem.status,
        "expected_return": expected_return,
        "expected_vol":    expected_vol,
        "sharpe_est":      expected_return / max(expected_vol, 1e-8),
        "n_long":          sum(1 for v in weights.values() if v > 0),
        "n_short":         sum(1 for v in weights.values() if v < 0),
    }

    return weights, meta


# ── EFFICIENT FRONTIER ────────────────────────────────────────────────────────

def compute_efficient_frontier(
    factor_scores: pd.Series,
    returns_history: pd.DataFrame,
    long_tickers: List[str],
    short_tickers: List[str],
    confidence_levels: List[float] = None,
) -> pd.DataFrame:
    """
    Run the optimizer at multiple target return levels to trace the
    efficient frontier — exactly the Solver multi-confidence workflow
    from your finance grad school.

    Returns DataFrame with columns:
        target_return, achieved_return, achieved_vol, sharpe, weights_json

    This is a diagnostic tool — Phase 1 runs this on sample windows
    to show the tradeoff curve and validate the optimizer is working.
    """
    if confidence_levels is None:
        confidence_levels = [0.90, 0.93, 0.95, 0.97, 0.98, 0.99]

    results = []
    all_tickers = long_tickers + short_tickers

    # Get base covariance once
    valid_hist = returns_history[all_tickers].dropna()
    cov = ewma_covariance(valid_hist)
    mu = np.array([factor_scores.get(t, 0) * optimizer_cfg.target_return for t in all_tickers])

    for confidence in confidence_levels:
        # Interpret confidence as: achieve this percentile of the return distribution
        # Higher confidence = higher target return = higher risk required
        target_scalar = _confidence_to_target(confidence, mu, cov)
        weights, meta = optimize_portfolio(
            factor_scores=factor_scores,
            returns_history=returns_history,
            long_tickers=long_tickers,
            short_tickers=short_tickers,
        )

        results.append({
            "confidence":      confidence,
            "target_return":   target_scalar,
            "achieved_return": meta.get("expected_return", 0),
            "achieved_vol":    meta.get("expected_vol", 0),
            "sharpe":          meta.get("sharpe_est", 0),
            "solve_status":    meta.get("status", "unknown"),
        })

    return pd.DataFrame(results)


def _confidence_to_target(
    confidence: float,
    mu: np.ndarray,
    cov: np.ndarray,
) -> float:
    """
    Convert a confidence level to a target return.
    At 99% confidence, target = top 1% of achievable return.
    Uses the maximum Sharpe portfolio's return × confidence scalar.
    """
    max_achievable = float(np.max(mu))
    return max_achievable * confidence


# ── FALLBACK ──────────────────────────────────────────────────────────────────

def _equal_weight_fallback(
    long_tickers: List[str],
    short_tickers: List[str],
) -> Dict[str, float]:
    """
    Equal-weight fallback when QP solve fails.
    Used for robustness — the backtest should never crash on a bad window.
    """
    n_long  = len(long_tickers)
    n_short = len(short_tickers)
    long_ratio  = optimizer_cfg.long_ratio
    short_ratio = 1 - long_ratio

    weights = {}
    if n_long  > 0: weights.update({t: long_ratio  / n_long  for t in long_tickers})
    if n_short > 0: weights.update({t: -short_ratio / n_short for t in short_tickers})
    return weights

File: polyglot/optimizer_jl/qp_solver.jl
#=
blaque_baux/optimizer_jl/qp_solver.jl
──────────────────────────────────────
Portfolio optimizer using JuMP + Clarabel.

Replaces qp_solver.py (CVXPY) with identical problem formulation:

    minimize    w' Σ w  +  λ_turn ‖w - w_prev‖₁  +  λ_oracle ‖w - w_oracle‖²
    subject to:
        Σ w_long   = long_ratio
        Σ w_short  = -short_ratio
        w_long  ∈ [min_wt, max_wt]
        w_short ∈ [-max_wt, -min_wt]
        μ' w ≥ target_return - 2 × exec_drag
        μ' w ≥ -sVaR  (if sVaR constraint provided)

JuMP advantages over CVXPY:
  - Clarabel runs natively (no Python→C→Python round-trip)
  - JuMP's constraint DSL is more expressive for adding lane variants
  - Julia's JIT compiles the model construction itself, not just the solve
  - Efficient frontier loop amortizes model construction across solves

Called from Python via juliacall:
    from juliacall import Main as jl
    jl.include("optimizer_jl/qp_solver.jl")
    weights, meta = jl.optimize_portfolio(factor_scores, returns_hist, ...)
=#

module QPSolver

using JuMP
using Clarabel
using LinearAlgebra
using Statistics

export optimize_portfolio, compute_efficient_frontier, ewma_covariance

# ── Configuration (mirrors config.OptimizerConfig) ───────────────────────────

Base.@kwdef struct OptimizerConfig
    long_ratio::Float64      = 0.50
    max_position_wt::Float64 = 0.08
    min_position_wt::Float64 = 0.01
    target_return::Float64   = 0.008
    turnover_penalty::Float64 = 0.0005
    risk_aversion::Float64   = 0.5
    execution_drag::Float64  = 0.0008
end

const DEFAULT_CONFIG = OptimizerConfig()


# ── EWMA Covariance ──────────────────────────────────────────────────────────

"""
    ewma_covariance(returns; λ=0.94, min_periods=20)

RiskMetrics EWMA covariance matrix. λ=0.94 for 15-min bars.
Returns (n×n) positive semi-definite matrix.
"""
function ewma_covariance(returns::Matrix{Float64}; λ::Float64=0.94, min_periods::Int=20)
    T, n = size(returns)

    if T < min_periods
        # Diagonal fallback
        return Diagonal(vec(var(returns, dims=1)))
    end

    # EWMA weights (most recent = highest weight)
    α = 1.0 - λ
    weights = [α * λ^(T - t) for t in 1:T]
    weights ./= sum(weights)

    # Weighted covariance
    μ = vec(sum(returns .* weights, dims=1))
    centered = returns .- μ'
    Σ = zeros(n, n)
    @inbounds for t in 1:T
        x = @view centered[t, :]
        Σ .+= weights[t] .* (x * x')
    end

    # Ensure PSD (clip negative eigenvalues)
    return ensure_psd(Σ)
end

function ensure_psd(M::Matrix{Float64}; ε::Float64=1e-6)
    F = eigen(Symmetric(M))
    vals = max.(F.values, ε)
    return F.vectors * Diagonal(vals) * F.vectors'
end


# ── Risk parameters (from risk_engine.py) ────────────────────────────────────

Base.@kwdef struct RiskParams
    long_ratio::Float64       = 0.50
    short_ratio::Float64      = 0.50
    var_constraint::Union{Float64, Nothing}  = nothing
    svar_constraint::Union{Float64, Nothing} = nothing
    position_scalar::Float64  = 1.0
    trend_label::String       = "neutral"
    trend_strength::Float64   = 0.0
end


# ── Main optimizer ───────────────────────────────────────────────────────────

"""
    optimize_portfolio(factor_scores, returns_history, long_idx, short_idx; kwargs...)

Main QP solve for one 15-minute window.

# Arguments
- `factor_scores::Vector{Float64}`: expected relative returns per asset
- `returns_history::Matrix{Float64}`: (T × n_total) returns for covariance
- `long_idx::Vector{Int}`: indices of long candidates (1-based)
- `short_idx::Vector{Int}`: indices of short candidates (1-based)
- `prev_weights::Vector{Float64}`: previous window's weights (for turnover)
- `borrow_costs::Vector{Float64}`: annualized borrow cost per asset
- `risk_params::RiskParams`: VaR/sVaR constraints from risk engine
- `oracle_prior::Union{Nothing, Vector{Float64}}`: L4 oracle weights
- `oracle_blend::Float64`: oracle regularization coefficient
- `cfg::OptimizerConfig`: optimizer hyperparameters

# Returns
- `weights::Vector{Float64}`: optimal weights (positive=long, negative=short)
- `meta::Dict{String,Any}`: diagnostics
"""
function optimize_portfolio(
    factor_scores::Vector{Float64},
    returns_history::Matrix{Float64},
    long_idx::Vector{Int},
    short_idx::Vector{Int};
    prev_weights::Union{Nothing, Vector{Float64}} = nothing,
    borrow_costs::Union{Nothing, Vector{Float64}} = nothing,
    risk_params::RiskParams = RiskParams(),
    oracle_prior::Union{Nothing, Vector{Float64}} = nothing,
    oracle_blend::Float64 = 0.0,
    cfg::OptimizerConfig = DEFAULT_CONFIG,
)
    n_long  = length(long_idx)
    n_short = length(short_idx)
    n_total = n_long + n_short
    all_idx = vcat(long_idx, short_idx)

    if n_total < 4
        return equal_weight_fallback(n_long, n_short, cfg),
               Dict{String,Any}("status" => "fallback")
    end

    # ── Expected returns ─────────────────────────────────────────────────
    μ = factor_scores[all_idx] .* cfg.target_return

    # Adjust for borrow costs on short positions
    if borrow_costs !== nothing
        for (i, idx) in enumerate(short_idx)
            local_i = n_long + i
            cost_per_window = borrow_costs[idx] / (252 * 26)
            μ[local_i] -= cost_per_window
        end
    end

    # ── Covariance matrix ────────────────────────────────────────────────
    Σ = ewma_covariance(returns_history[:, all_idx])

    # ── Dynamic L/S ratio ────────────────────────────────────────────────
    lr = risk_params.long_ratio
    sr = risk_params.short_ratio

    # ── Build JuMP model ─────────────────────────────────────────────────
    model = Model(Clarabel.Optimizer)
    set_silent(model)
    set_optimizer_attribute(model, "max_iter", 200)

    @variable(model, w[1:n_total])

    # ── Objective: risk_aversion × w'Σw + turnover + oracle prior ────────
    @objective(model, Min,
        cfg.risk_aversion * w' * Σ * w
    )

    # Turnover penalty (L1 via auxiliary variables)
    if prev_weights !== nothing
        w_prev = prev_weights[all_idx]
        @variable(model, t_abs[1:n_total] >= 0)
        @constraint(model, [i=1:n_total],  w[i] - w_prev[i] <= t_abs[i])
        @constraint(model, [i=1:n_total], -w[i] + w_prev[i] <= t_abs[i])

        # Add to objective
        set_objective_function(model,
            objective_function(model) + cfg.turnover_penalty * sum(t_abs)
        )
    end

    # Oracle prior: L2 regularization toward oracle weights (L4 lane)
    if oracle_prior !== nothing && oracle_blend > 0
        w_oracle = oracle_prior[all_idx]
        set_objective_function(model,
            objective_function(model) + oracle_blend * sum((w[i] - w_oracle[i])^2 for i in 1:n_total)
        )
    end

    # ── Constraints ──────────────────────────────────────────────────────

    # Long book: sum to long_ratio, bounded per position
    @constraint(model, sum(w[1:n_long]) == lr)
    @constraint(model, [i=1:n_long], w[i] >= cfg.min_position_wt)
    @constraint(model, [i=1:n_long], w[i] <= cfg.max_position_wt)

    # Short book: sum to -short_ratio, bounded per position
    @constraint(model, sum(w[n_long+1:n_total]) == -sr)
    @constraint(model, [i=n_long+1:n_total], w[i] <= -cfg.min_position_wt)
    @constraint(model, [i=n_long+1:n_total], w[i] >= -cfg.max_position_wt)

    # Minimum expected return
    @constraint(model, μ' * w >= cfg.target_return - 2 * cfg.execution_drag)

    # sVaR constraint (from risk engine)
    if risk_params.svar_constraint !== nothing
        svar = risk_params.svar_constraint
        if svar > 0 && !isnan(svar)
            @constraint(model, μ' * w >= -svar)
        end
    end

    # ── Solve ────────────────────────────────────────────────────────────
    t0 = time()
    optimize!(model)
    solve_ms = (time() - t0) * 1000

    status = string(termination_status(model))

    if !(termination_status(model) in [MOI.OPTIMAL, MOI.ALMOST_OPTIMAL])
        return equal_weight_fallback(n_long, n_short, cfg),
               Dict{String,Any}(
                   "status" => status,
                   "solve_ms" => solve_ms,
                   "long_ratio" => lr,
                   "short_ratio" => sr,
               )
    end

    # ── Package results ──────────────────────────────────────────────────
    w_val = value.(w)

    # Clean up noise
    for i in 1:n_total
        if abs(w_val[i]) < cfg.min_position_wt * 0.5
            w_val[i] = 0.0
        end
    end

    # Diagnostics
    exp_ret = μ' * w_val
    exp_vol = sqrt(max(w_val' * Σ * w_val, 0.0))

    meta = Dict{String,Any}(
        "status"          => status,
        "solve_ms"        => solve_ms,
        "expected_return" => exp_ret,
        "expected_vol"    => exp_vol,
        "sharpe_est"      => exp_ret / max(exp_vol, 1e-8),
        "n_long"          => count(x -> x > 0, w_val),
        "n_short"         => count(x -> x < 0, w_val),
        "long_ratio"      => lr,
        "short_ratio"     => sr,
        "var_constraint"  => risk_params.var_constraint,
        "svar_constraint" => risk_params.svar_constraint,
        "trend_label"     => risk_params.trend_label,
        "trend_strength"  => risk_params.trend_strength,
        "position_scalar" => risk_params.position_scalar,
        "oracle_active"   => oracle_prior !== nothing,
    )

    return w_val, meta
end


# ── Efficient frontier ───────────────────────────────────────────────────────

"""
    compute_efficient_frontier(factor_scores, returns_history, long_idx, short_idx;
                               confidence_levels=[0.90,0.93,0.95,0.97,0.98,0.99])

Trace the efficient frontier at multiple target return levels.
Exactly the Solver multi-confidence workflow.

Returns Vector of (confidence, target_return, achieved_return, achieved_vol, sharpe, status).
"""
function compute_efficient_frontier(
    factor_scores::Vector{Float64},
    returns_history::Matrix{Float64},
    long_idx::Vector{Int},
    short_idx::Vector{Int};
    confidence_levels::Vector{Float64} = [0.90, 0.93, 0.95, 0.97, 0.98, 0.99],
    cfg::OptimizerConfig = DEFAULT_CONFIG,
)
    results = Vector{Dict{String,Any}}()

    for conf in confidence_levels
        # Scale target return by confidence
        max_achievable = maximum(factor_scores[vcat(long_idx, short_idx)])
        target = max_achievable * conf

        local_cfg = OptimizerConfig(
            long_ratio      = cfg.long_ratio,
            max_position_wt = cfg.max_position_wt,
            min_position_wt = cfg.min_position_wt,
            target_return   = target,
            turnover_penalty = cfg.turnover_penalty,
            risk_aversion   = cfg.risk_aversion,
            execution_drag  = cfg.execution_drag,
        )

        _, meta = optimize_portfolio(
            factor_scores, returns_history,
            long_idx, short_idx;
            cfg=local_cfg,
        )

        push!(results, Dict{String,Any}(
            "confidence"      => conf,
            "target_return"   => target,
            "achieved_return" => get(meta, "expected_return", 0.0),
            "achieved_vol"    => get(meta, "expected_vol", 0.0),
            "sharpe"          => get(meta, "sharpe_est", 0.0),
            "status"          => get(meta, "status", "unknown"),
        ))
    end

    return results
end


# ── Equal weight fallback ────────────────────────────────────────────────────

function equal_weight_fallback(n_long::Int, n_short::Int, cfg::OptimizerConfig)
    n_total = n_long + n_short
    w = zeros(n_total)
    if n_long > 0
        w[1:n_long] .= cfg.long_ratio / n_long
    end
    if n_short > 0
        w[n_long+1:end] .= -(1.0 - cfg.long_ratio) / n_short
    end
    return w
end

end # module QPSolver
